ops/model.py

```diff
@@ -1478,7 +1478,7 @@ def __init__(
             backend: '_ModelBackend', cache: '_ModelCache', active: bool = True):
         self.name = relation_name
         self.id = relation_id
-        self.app: Optional[Application] = None
+        self.app: Application = typing.cast(Application, None)
```

Comment:
Hmm, this gets a bit ugly, since we're lying about the type. What about having a local `app` variable (with an implicit type of `Application|None`) and then a `self.app = app` at the end? Probably with an `assert app is not None` first to make type checkers happy.

Reply:
After some discussion and according to @tonyandrewmeyer and @benhoyt:
- `self.app` will always be set because we have a fallback mechanism to get it from env vars and stuff.
- In case of a weird Juju bug or edge case where the app is actually None, we probably do not want the charm to raise an AssertionError.

So, we will use typing.cast and add a comment to explain that it will not be None.